Adds ATHS support

conorsch opened the pull request:
See rationale in #12. Succinctly, enables an extra layer of authentication for hidden services intended to have only selective availability, e.g. SSH. Clients will need to update their torrc files with the corresponding `HidServAuth` line in order to connect to the ATHS. (It's not right for everything, clearly, but it's very nice to have for SSH in particular.)

To enable it, add a list var `hidden_service_authorized_clients` to the services configuration:

```
hidden_service_services:
  ssh:
    hidden_service_ports:
      - [22, 22]
    hidden_service_authorized_clients:
      - admin-{{ inventory_hostname }}
```

Plays nicely with #11, in that the `hostname` file is still displayed in its entirety, including:
- the Onion URL
- the auth cookie that needs to go in the client torrc
- a comment with name of the authorized client

Appending inventory_hostname in the authorized_client var causes the comment in the hostname config to render as e.g. `# client: admin-www.example.com`, which greatly helps when managing auth cookies for multiple hosts.

Closes #12.

conorsch committed:
Adds support for Authenticated Tor Hidden Services

The ATHS requires connecting clients to provide an auth cookie,
specified as HidServAuth in the client torrc, in order to communicate
with the hidden service. Specifying a hidden service authenticated in
this role now requires an additional key in the service config dict.

The new functionality is off by default. The logic displaying hostnames
still displays the HidServAuth info, but the formatting could be better.

conorsch committed:
Special handling for HidServAuth config line

The HidServAuth line can be repeated, and must be, so assume that vars
for this config option are a list, and iterate over them. Haven't
searched the tor manpage for other listable vars; in the future it may
be convenient to store a role default var with a list of listvars,
and update the template to reference that.

conorsch committed:
Updates docs with authorized client example vars

Nothing too egregious, just two entries showing how it's done. More
explicit recommendations can be made in the README, but I haven't taken
the liberty.

t2d commented:
Can you please explain `HiddenServiceAuthorizeClient`? As far as I understand it, this is a list of authorized host names. Docs say "Valid client names are 1 to 16 characters long and only use characters in A-Za-z0-9+-_ (no spaces)". This seams to collide with your default `admin-{{ inventory_hostname }}`, no?

Can you explain the workflow with these options?
First, I add `admin-{{ inventory_hostname }}` to `HiddenServiceAuthorizeClient` in the server variables. Than I copy what is generated in a run under  `# client: admin-www.example.com` and add it to the clients var at `HidServAuth`. Correct?

Shouldn't we include `HidServerAuth` in the variables and example in README also?

conorsch commented:
> This seams to collide with your default admin-{{ inventory_hostname }}, no?

For shame, I hadn't adequately tested that. In practice I've been manually maintaining comments in an encrypted vars file, and while documenting the functionality it occurred to me to leverage the name. Sounds like the safest course is just to advise folks to make note of which values map to which hosts.

> Can you explain the workflow with these options?
1. Include the `hidden_service_authorized_clients` list as a subelement of the hidden service dict, e.g.
   
   ```
   hidden_service_services:
   ssh:
     hidden_service_ports:
       - [22, 22]
     hidden_service_authorized_clients:
       - admin
   ```
2. Run the role over standard IPv4 to generate hidden service hostname and private key.
3. Copy the hostname info from the debug output (added in #11) and paste into the client torrc.
4. Reload client tor service.
5. Rerun playbook over tor to ensure idempotence.

Step 3 could provide more explicit instructions specific to ATHS config, since the contents of the `hostname` file will obviously be for either a THS or an ATHS (a THS will have only the Onion URL and nothing else). 

> Shouldn't we include HidServerAuth in the variables and example in README also?

In general I'm always happy to add documentation. In real-world use cases, I expect there to be a bit of tooling necessary for managing (A)THS values for SSH, tooling that this role can't realistically provide. For example, [here's a playbook](https://gist.github.com/conorsch/85d568d2c13e1da98c43cbcbf391b62c) that I use that should give a rough idea of the provision -> fetch -> lookup workflow. There's also a corresponding dynamic inventory script. Leaving all that aside for now, I'll mention just the `HidServAuth` option in the README and hopefully that will be sufficient.

xshadow commented:
Thanks @conorsch for bringing that up. 

I think thats a good feature and I'm willing to merge that. 

But I assume,  there is something missing: 

The variable

> HidServAuth 

is missing in the default vars file and should get the correct naming.
As far as I understood the docs correct, it can be found in the hostname file, such as the hostname. 

Additionally, the default value for the list should be in the template, but also in the defaults vars file. 

> {% for client in hsproperties['hidden_service_authorized_clients']|default([]) %}

t2d commented:
@xshadow I believe `HidServAuth` won't be there on default.

```
Additionally, the default value for the list should be in the template, but also in the defaults vars file.
{% for client in hsproperties['hidden_service_authorized_clients']|default([]) %}
```

While this is indeed a good addition, I believe we could merge without it also.

xshadow commented:
@t2d yes you are right that it not has to be there on default.

@conorsch 

The following corrections are needed for a merge:
1. in Readme - this should be corrected

```
+    hidden_service_authorized_clients:
+      - admin-{{ inventory_hostname }} 
```
1. The var `HidServAuth` --> should be renamed to `hidden_service_hid_serv_auth` and added as a possible option in the Readme
2. It would be great if you could add the default value [] for the authorized_client list also in to  the defaults vars file.
   
   > {% for client in hsproperties['hidden_service_authorized_clients']|default([]) %}

conorsch committed:
Removes recommendation to use "inventory_hostname" in auth clients list

The README contained a misguided recommedation to concatenate the hidden
service client name with the inventory_hostname, for readability and
maintenance across many hosts. Unfortunately it violates the formatting
permitted by Tor, and therefore can't be used. (This is why you test
things before recommending them.)

conorsch committed:
Documents default authorized clients list (empty) in default vars

Preserving the |default([]) in the template because users of the role
may override the services config dict and omit the authorized clients
key. In that case, we still want the torrc template to be written, just
without and HidServAuth related lines.

conorsch committed:
Fixes undefined handling of hostname for hidden service

The "when" conditional is checking a dict key, and Ansible will throw an
error about an undefined value if it's not present in the dict. Easy to
sidestep with |default(false), which matches the format of the other
when conditionals in the task list.

conorsch commented on defaults/main.yml:
```diff
@@ -8,8 +8,8 @@ hidden_service_services:
 hidden_services_configuration:
   SocksPort: 9050
   SocksPolicy: "reject *"
-
```
Automatic linting, I can revert if you feel strongly about trailing empty lines (and want them). :smile:

conorsch commented on tasks/main.yml:
```diff
@@ -90,7 +90,7 @@
   changed_when: false
   with_dict: "{{ hidden_service_services }}"
-  when: not item.value.hidden_service_hostname
+  when: not item.value.hidden_service_hostname|default(false)
```
Adding the default catch-all for consistency with the other "when" conditionals.

conorsch commented on templates/torrc.j2:
```diff
@@ -60,7 +60,13 @@
 #CookieAuthentication 1
 
 {% for key, value in hidden_services_configuration.iteritems() %}
+{% if key == 'HidServAuth' %}
```
There are multiple values that can be repeated in the torrc—consider adding a list var that stores the names of those options, so _all_ list-friendly config options can be iterated over. Not adding in this PR because it seems out of scope. For now, the only option handled as a list rather than a string is `HidServAuth`.

conorsch committed:
Converts HidServAuth config values to list var

Using a top-level default var, called `hidden_service_hid_serv_auth`,
that expects a list and writes `HidServAuth` lines to the torrc.

Updates docuementation in README accordingly.

conorsch commented on README.md:
```diff
@@ -95,6 +103,13 @@ hidden_services_configuration:
+# hostname file and add them to the torrc for intended clients, e.g.
+# the Ansible controller, via the list var below.
+hidden_service_hid_serv_auth:
+  - "r7w3xdf3r5smxokv.onion p0xMVci7ffeQFA4IWkcBxR # client: admin"
```
ATHS value generated specifically for README; the host has since been destroyed, so it's safe to document here.

conorsch commented:
@xshadow I've updated the README and default vars as you requested. Please take another look. Happy to rebase/squash after review.

xshadow commented:
@conorsch that looks good. I'm willing to merge it. Nothing changes in our configuration

Are you sure that this option is needed. From my point of view it looks like HidServAuth is part of the client torrc.

> HidServAuth onion-address auth-cookie [service-name]
>     Client authorization for a hidden service. Valid onion addresses contain 16 characters in a-z2-7 plus ".onion", and valid auth cookies contain 22 characters in A-Za-z0-9+/. The service name is only used for internal purposes, e.g., for Tor controllers. This option may be used multiple times for different hidden services. If a hidden service uses authorization and this option is not set, the hidden service is not accessible. Hidden services can be configured to require authorization using the HiddenServiceAuthorizeClient option.

I think on the server side only the:

> HiddenServiceAuthorizeClient auth-type client-name,client-name,…

is necessary.  Else we would configure it somehow twice and that doesn't make sense to me.

If so, that's not needed in the Readme:

> +# Hosts that specified `hidden_service_authorized_clients` will generate
> +# auth cookies for restricted access. Collect those values from the
> +# hostname file and add them to the torrc for intended clients, e.g.
> +# the Ansible controller, via the list var below.
> +hidden_service_hid_serv_auth:
> -  - "r7w3xdf3r5smxokv.onion p0xMVci7ffeQFA4IWkcBxR # client: admin"

instead there should be 

> hidden_service_services:
>   ssh:
>     hidden_service_hostname:
>     hidden_service_ports:
>       - [22, 22]
>     hidden_service_authorized_clients: []
>     # Hosts that specified `hidden_service_authorized_clients` will generate
>    # auth cookies for restricted access. Collect those values from the 
>    # hostname file and add them to the torrc of the clients
>    # such as:   HidServAuth onion-address auth-cookie

t2d commented:
@xshadow I understood, that @conorsch wants to configure both client and server with this role. And yes, each party uses only one of these options. @conorsch Is this correct?

conorsch commented:
Thanks for chiming in, @t2d. Yes, you are correct: I'm using this role for both client and server configuration. Doing so allows me to use a vars-based override for the Ansible controller, so that the HidServAuth values from remote hosts are populated in my local workstation's torrc file. Doing so allows me to connect via SSH to all clients over ATHS, and onboarding newly created hosts is as simple as rerunning the playbook.

I see there are now conflicts with master—I'll look into the changes and resolve.

conorsch commented:
Conflicts resolved—please take another look.

xshadow commented:
Thank @conorsch , sorry for taking that long.

conorsch commented:
@xshadow Thanks for the thorough review! :+1:
